components/pd_client/src/client.rs

```diff
@@ -46,12 +48,49 @@ impl RpcClient {
             v => v.checked_add(1).unwrap_or(std::isize::MAX),
         };
         for i in 0..retries {
-            match validate_endpoints(Arc::clone(&env), cfg, &security_mgr) {
+            match validate_endpoints(Arc::clone(&env), cfg, security_mgr.clone()) {
                 Ok((client, members)) => {
-                    return Ok(RpcClient {
+                    let rpc_client = RpcClient {
                         cluster_id: members.get_header().get_cluster_id(),
-                        leader_client: LeaderClient::new(env, security_mgr, client, members),
-                    });
+                        leader_client: Arc::new(LeaderClient::new(
+                            env,
+                            security_mgr,
+                            client,
+                            members,
+                        )),
+                    };
+
+                    // spawn a background future to update PD information periodically
+                    let duration = Duration::from_secs(cfg.update_interval);
+                    let update_loop =
+                        loop_fn(Arc::downgrade(&rpc_client.leader_client), move |client| {
+                            GLOBAL_TIMER_HANDLE
+                                .delay(Instant::now() + duration)
+                                .then(|_| {
+                                    match client.upgrade() {
+                                        Some(cli) => {
+                                            let req = cli.reconnect();
+                                            future::Either::A(req.then(|r| {
+                                                if r.is_err() {
+                                                    warn!("update PD information failed");
+                                                    // will update later anyway
+                                                }
+                                                Ok(Loop::Continue(client))
+                                            }))
+                                        }
+                                        // if the client has been dropped, we can stop
+                                        None => future::Either::B(ok(Loop::Break(()))),
+                                    }
+                                })
+                        });
+                    rpc_client
+                        .leader_client
+                        .inner
+                        .rl()
+                        .client_stub
+                        .spawn(update_loop);
```

Comment:
I already encountered this problem when testing this PR. The root cause of #7124 is that `reconnect` is blocking. When we spawn a future that contains a call to `reconnect`, it may block the executor. Currently, the problem is mitigated by not spawning any future onto the GRPC executor. However, it is still possible that these futures may block other executors. I think the best solution is to make `reconnect` non-blocking, as in this PR.